src/UnwindRustSgx.c

```diff
@@ -0,0 +1,150 @@
+#include "UnwindRustSgx.h"
+
+#ifdef __cplusplus
+extern "C"  {
+#endif
+	int c_rwlock_rdlock(RWLock *rwlock);
+	int c_rwlock_wrlock(RWLock *rwlock);
+	int c_rwlock_unlock(RWLock *rwlock);
+	unsigned char *__rust_alloc(size_t, size_t);
+	void __rust_dealloc(unsigned char *, size_t);
+	int c_print_msg(uint8_t *m, int s);
+	void c_abort(void);
+	unsigned char *c_encl_address(size_t);
+
+#ifdef __cplusplus
+}
+#endif
+
+void abort() __attribute__((noreturn)) {
+	c_abort();
+}
+
+static int vwrite_err(const char *format, va_list ap)
+{
+	char s[max_log];
+	int len = 0;
+	s[0]='\0';
+	len = vsnprintf(s, max_log, format, ap);
+	return c_print_msg((uint8_t *)s, len);
+}
+
+static int write_err(const char *format, ...)
+{
+	int ret;
+	va_list args;
+        va_start(args, format);
+	ret = vwrite_err(format, args);
+	va_end(args);
+
+
+	return ret;
+}
+
+int libuw_fprintf(FILE *stream, const char *format, ...)
+{
+
+	int ret;
+	if (stream != stderr) {
+		write_err("Rust SGX Unwind supports only writing to stderr\n");
+		return -1;
+	} else {
+		va_list args;
+		int ret = 0;
+		va_start(args, format);
+		ret += vwrite_err(format, args);
+		va_end(args);
+	}
+
+	return ret;
+}
+
+
+void __libuw_assert_fail(const char * assertion, const char * file, unsigned int line, const char * function)
+{
+	write_err("%s:%d %s %s\n", file, line, function, assertion);
+	abort();
+}
+
+
+
+int libuw_pthread_rwlock_rdlock(RWLock *rwlock) {
+	int ret = 0;
+	ret = c_rwlock_rdlock(rwlock);
+	return ret;
+}
+
+int libuw_pthread_rwlock_wrlock(RWLock *rwlock) {
+	return c_rwlock_wrlock(rwlock);
+}
+
+int libuw_pthread_rwlock_unlock(RWLock *rwlock) {
+	return c_rwlock_unlock(rwlock);
+}
+
+
+// We do not report stack over flow detected.
+// Calling write_err uses more stack due to the way we have implemented it.
+// With possible enabling of stack probes, we should not
+// get into __stack_chk_fail() at all.
+void __libuw_stack_chk_fail() {
+	abort();
+}
+
+extern uint64_t ENCLAVE_TEXT_BASE;
+extern uint64_t ENCLAVE_TEXT_SIZE;
+extern uint64_t EH_FRM_HDR_BASE;
+extern uint64_t EH_FRM_HDR_SIZE;
+
+typedef Elf64_Phdr Elf_Phdr; 
+int
+dl_iterate_phdr (int (*callback) (struct dl_phdr_info *,
+				  size_t, void *),
+		 void *data)
+{
+	struct dl_phdr_info info;
+	struct dl_phdr_info *pinfo = &info;
+	int ret = 0;
+
+        Elf_Phdr phdr[2];
+
+	size_t text_size = ENCLAVE_TEXT_SIZE;
+	size_t eh_base_size = EH_FRM_HDR_SIZE;
+
+	memset(pinfo, 0, sizeof(*pinfo));
+
+	pinfo->dlpi_addr = (void *)c_encl_address(0);
+	pinfo->dlpi_phnum = 2;
+
+	pinfo->dlpi_phdr = phdr;
+	memset(phdr, 0, 2*sizeof(*phdr));
+
+
+	phdr[0].p_type = PT_LOAD;
+	phdr[0].p_vaddr = (size_t)ENCLAVE_TEXT_BASE;
+	phdr[0].p_memsz = text_size;
+
+	phdr[1].p_type = PT_GNU_EH_FRAME;
+	phdr[1].p_vaddr = (size_t)EH_FRM_HDR_BASE;
+	phdr[1].p_memsz = eh_base_size;
+
+
+	ret = callback (&info, sizeof (struct dl_phdr_info), data);
+	return ret;
+}
+
+void *libuw_malloc(size_t size)
+{
+	return (void *)__rust_alloc(size, 1);
```

Comment:
`__rust_dealloc` requires that you pass the size. `malloc` normally uses a minimum alignent of `sizeof(size_t)`